File: Baseline/matlab/runSparseDNNchallenge.m
% Set locations of files.
inputFile = '../../code/data/MNIST/sparse-images-';
categoryFile = '../../code/data/DNN/neuron';
layerFile = '../../code/data/DNN/neuron';

% Select DNN to run.
%Nneuron = [1024, 4096, 16384, 65536];
Nneuron = [1024];
SAVECAT = 0;    % Overwrite truth categories.
READTSV = 1;    % Read input and layers from .tsv files.
READMAT = 0;    % Redd input and layers from .mat files.

% Select number of layers to run.
%maxLayers = 120 * [1, 4, 16];
maxLayers = [120];

% Set DNN bias.
neuralNetBias = [-0.3,-0.35,-0.4,-0.45];

% Loop over each DNN.
for i=1:length(Nneuron)

  % Load sparse MNIST data.
  if READTSV
    featureVectors = readTriples([inputFile num2str(Nneuron(i)) '.tsv']);
  end
  if READMAT
    load(['data/MNIST/sparse-images-' num2str(Nneuron(i)) '.mat'],'z');
    featureVectors = z;
  end

  featureVectors(end,Nneuron(i)) = 0;       % Pad matrix.
  NfeatureVectors = size(featureVectors,1);
  
  % Read layers.
  for j=1:length(maxLayers)
  
    % Read in true categories.
    if not(SAVECAT)
      trueCategories = str2num(StrFileRead([categoryFile num2str(Nneuron(i)) '-l' num2str(maxLayers(j)) '-categories.tsv']));
    end

    DNNedges = 0;
    layers = {};
    bias = {};
    tic;
      for k=1:maxLayers(j)
        if READTSV
          layers{k} = readTriples([layerFile num2str(Nneuron(i)) '/n' num2str(Nneuron(i)) '-l' num2str(k) '.tsv']);
        end
        if READMAT
          load([layerFile num2str(Nneuron(i)) '/n' num2str(Nneuron(i)) '-l' num2str(k) '.mat'],'layersScaledj');
          layers{k} = layersScaledj;
        end
        DNNedges = DNNedges + nnz(layers{k});
        bias{k} = sparse(ones(1,Nneuron(i)) .* neuralNetBias(i));
      end
    readLayerTime = toc;
    readLayerRate = DNNedges/readLayerTime;

    disp(['DNN neurons/layer: ' num2str(Nneuron(i)) ', layers: ' num2str(maxLayers(j)) ', edges: ' num2str(DNNedges)]);
    disp(['Read time (sec): ' num2str(readLayerTime) ', read rate (edges/sec): ' num2str(readLayerRate)]);

    % Perform and time challenge
    tic;
       scores = inferenceReLUvec(layers,bias,featureVectors);  
    challengeRunTime = toc;

    challengeRunRate = NfeatureVectors*DNNedges/challengeRunTime;
    disp(['Run time (sec): ' num2str(challengeRunTime) ', run rate (edges/sec): ' num2str(challengeRunRate)]);

    % Compute categories from scores.
    [categories col val] = find(sum(scores,2));

    if SAVECAT
      StrFileWrite(sprintf('%d\n',categories),[categoryFile num2str(Nneuron(i)) '-l' num2str(maxLayers(j)) '-categories.tsv']);
    else
      categoryDiff = sparse(trueCategories,1,1,NfeatureVectors,1) - sparse(categories,1,1,NfeatureVectors,1);
      if (nnz(categoryDiff))
        disp('Challenge FAILED');
      else
        disp('Challenge PASSED');
      end
    end
    
  end

end


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Software Engineer: Dr. Jeremy Kepner                    
% MIT                   
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% (c) <2019> Massachusetts Institute of Technology
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File: Baseline/C++/runSparseDNNchallenge.cpp
#include<iostream>
#include<string>
#include <chrono>
#include<Eigen/Sparse>
#include "./ReadCategories.cpp"
#include "./ReadSparse.cpp"
#include "./inferenceReLUvec.cpp"
#include "./StrFileRead.cpp"

const std::string INPUTFILE = "../MNIST/sparse-images-";
const std::string CATEGORYFILE = "../MNIST/neuron";
const std::string LAYERFILE = "../neuron";

//currently ignore SAVCAT and READMAT.
const bool SAVECAT = false;
const bool READTSV = true;
const bool READMAT = false;

//total num of MNIST input
const int numMNIST = 60000;

//Select DNN to run.
//Nneuron = [1024, 4096, 16384, 65536];
int Nneuron[] = {1024};

//Select number of layers to run.
//maxLayers = 120 * [1, 4, 16];
int maxLayers[] = {120};

//Set DNN bias
float neuralNetBias[] = {-0.3f, -0.35f, -0.4f, -0.45f};

int main(){
	size_t Nlen = sizeof(Nneuron)/sizeof(Nneuron[0]);
	size_t maxLayersLen = sizeof(maxLayers)/sizeof(maxLayers[0]);
	Eigen::SparseMatrix<float> featureVectors(numMNIST, 1);
	featureVectors.reserve(numMNIST/200);
	for(size_t i = 0; i < Nlen; ++i){
		if(READTSV) featureVectors = ReadSparse(StrFileRead(INPUTFILE + std::to_string(Nneuron[i]) + ".tsv"), Nneuron[i], numMNIST, true);
		//else if(READMAT)
		int NfeatureVectors = featureVectors.cols();

		//Read layers.
		for(size_t j = 0; j < maxLayersLen; ++j){
			Eigen::SparseVector<float> trueCategories;
			if (!SAVECAT) trueCategories = ReadCategories(StrFileRead(CATEGORYFILE + std::to_string(Nneuron[i]) + "-l" + std::to_string(maxLayers[j]) +  "-categories.tsv"), numMNIST);
			int DNNedges = 0;
			std::vector<Eigen::SparseMatrix<float> > layers;
			layers.reserve(maxLayers[j]);
			std::vector<float> bias;
			bias.reserve(maxLayers[j]);

			//Read Layer
			auto startLayer = std::chrono::high_resolution_clock::now();
			for(int k = 0; k < maxLayers[j]; ++k){
				if(READTSV) layers.push_back(ReadSparse(StrFileRead(LAYERFILE + std::to_string(Nneuron[i]) + "/n" + std::to_string(Nneuron[i]) + "-l" +  std::to_string(k + 1) + ".tsv"), Nneuron[i], numMNIST,  false));
				//else if(READMAT);
				DNNedges += layers[k].nonZeros();
				bias.push_back(neuralNetBias[i]);
			}
			auto stopLayer = std::chrono::high_resolution_clock::now();
			auto readLayerTime = std::chrono::duration_cast<std::chrono::seconds>(stopLayer - startLayer).count();
			auto readLayerRate = DNNedges / readLayerTime;
			std::cout << "DNN neurons/layer: " << Nneuron[i] << ", layers: " << maxLayers[j] << ", edges: " <<  DNNedges << std::endl;
			std::cout << "Read time (sec):" << readLayerTime << ", Read rate (edges/sec):" << readLayerRate << std::endl;

			//Perform and time challenge
			auto startRun = std::chrono::high_resolution_clock::now();
			Eigen::SparseMatrix<float> scores = inferenceReLUvec(layers, bias, featureVectors);
			auto stopRun = std::chrono::high_resolution_clock::now();
			auto challengeRunTime = std::chrono::duration_cast<std::chrono::seconds>(stopRun - startRun).count();
			auto challengeRunRate = NfeatureVectors * (DNNedges / challengeRunTime);
			std::cout << "Run time (sec):" << challengeRunTime << ", Run rate (edges/sec):" <<challengeRunRate<< std::endl;

			//compute categories from scores.
			Eigen::SparseVector<float> categories = Eigen::MatrixXf(scores).rowwise().sum().sparseView();
			categories = categories.unaryExpr([] (float a) {
					if(a > 0) return 1;
					else return 0;
					});
			if (SAVECAT);
			else{
				Eigen::SparseVector<float> categoryDIff = trueCategories - categories;
				categoryDIff = categoryDIff.pruned();
				if(categoryDIff.nonZeros())
					std::cout << "Challenge FAILED";
				else
					std::cout << "Challenge PASSED" ;
				std::cout << std::endl;
			}
		}
	}
}
